src/main/daml/Daml/Finance/Interface/Lifecycle/Rule/Lifecycle.daml

```diff
@@ -37,21 +38,21 @@ interface Lifecycle where
     do
       pure $ view this
 
-  nonconsuming choice Evolve : (ContractId Lifecycle, [ContractId Effect.I])
+  nonconsuming choice Evolve : (ContractId Instrument.I, [ContractId Effect.I])
     -- ^ Process an event. It returns a tuple of the lifecycled instrument (or the original
     --   instrument when the former does not exist) and the effects.
     with
-      ruleName : Text
-        -- ^ The lifecycle rule to be processed.
       eventCid : ContractId Event.I
         -- ^ The event.
+      instrumentCid : ContractId Instrument.I
```

Comment:
I added an `instrumentCid` as we discussed yesterday. It however feels more natural to use `InstrumentKey` instead.

By using a key, client code doesn't have to go from `Key` to `ContractId` using the `Reference` workaround.

Happy to hear your thoughts

Reply:
Good point about authorization requirements. 

With `ContractId`, one needs to
- have visibility on the instrument contract (by being stakeholder or explicit disclosure)
- have authorization to `fetch` (by being stakeholder)

With `InstrumentKey`, one needs to
- have visibility on the `Reference` contract (by being stakeholder or explicit disclosure)
- have visibility on the corresponding instrument cid + authorization to `fetch` it (so being a stakeholder)

Once we get rid of the workaround, if we assume that `fetchByKey` for interfaces has the same authorization as the current `fetchByKey`, then the submitting party needs to be a stakeholder on the contract

I guess we can assume that the authorization requirements are indeed the same, so I will go for `InstrumentKey`.